src/invoice/verification.rs

```diff
@@ -0,0 +1,389 @@
+impl VerificationStrategy {
+    fn verify_signature(&self, sig: &Signature, cleartext: &[u8]) -> Result<(), SignatureError> {
+        let pk = base64::decode(sig.key.as_bytes())
+            .map_err(|_| SignatureError::CorruptKey(sig.key.clone()))?;
+        let sig_block = base64::decode(sig.signature.as_bytes())
+            .map_err(|_| SignatureError::CorruptSignature(sig.key.clone()))?;
+
+        let pubkey =
+            PublicKey::from_bytes(&pk).map_err(|_| SignatureError::CorruptKey(sig.key.clone()))?;
+        let ed_sig = EdSignature::new(
+            sig_block
+                .as_slice()
+                .try_into()
+                .map_err(|_| SignatureError::CorruptSignature(sig.key.clone()))?,
+        );
+        pubkey
+            .verify_strict(cleartext, &ed_sig)
+            .map_err(|_| SignatureError::Unverified(sig.key.clone()))
+    }
+    /// Verify that every signature on this invoice is correct.
+    ///
+    /// The verification strategy will determine how this verification is performed.
+    /// Depending on the selected strategy, the `[[signature]]` blocks will be evaluated
+    /// for the following:
+    ///
+    /// - Is the key in the keyring?
+    /// - Can the signature be verified?
+    ///
+    /// Note that the purpose of the keyring is to ensure that we know about the
+    /// entity that claims to have signed the invoice.
+    ///
+    /// If no signatures are on the invoice, this will succeed.
+    ///
+    /// A strategy will determine success or failure based on whether the signature is verified,
+    /// whether the keys are known, whether the requisite number/roles are satisfied, and
+    /// so on.
+    pub fn verify(&self, inv: &Invoice, keyring: &KeyRing) -> Result<(), SignatureError> {
+        let (roles, all_valid, all_verified, all_roles) = match self {
+            VerificationStrategy::GreedyVerification => {
+                (vec![SignatureRole::Creator], true, true, true)
+            }
+            VerificationStrategy::CreativeIntegrity => {
+                (vec![SignatureRole::Creator], false, true, true)
+            }
+            VerificationStrategy::AuthoritativeIntegrity => (
+                vec![SignatureRole::Creator, SignatureRole::Approver],
+                false,
+                false,
+                false,
+            ),
+            VerificationStrategy::ExhaustiveVerification => (
+                vec![
+                    SignatureRole::Creator,
+                    SignatureRole::Approver,
+                    SignatureRole::Host,
+                    SignatureRole::Proxy,
+                ],
+                true,
+                true,
+                false,
+            ),
+            VerificationStrategy::MultipleAttestation(a, b) => (a.clone(), b.clone(), true, true),
+        };
+
+        // Either the Creator or an Approver must be in the keyring
+        match inv.signature.as_ref() {
+            None => Ok(()),
```

Comment:
Should we at least log a warning if there are no signatures? Or is this going to go away once we fully enforce things?

Reply:
I will log an info for now. But, yeah, eventually the server will not accept invoices that do not have signatures. This PR just does the host signature. A different one will have to do the creator signature.